.github/workflows/Bash.yml

```
@@ -0,0 +1,69 @@
+# Originally created by Zernit project <https://github.com/RXT0112/Zernit/blob/master/.github/workflows/bash.yml>, granted to kiwibrowser under their preffered license.
+
+name: Bash
+
+# Relevant to events - https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows
+on: 
+  pull_request:
+    types: [synchronize, opened, reopened, ready_for_review]
+    paths:
+    - '**.bash'
+    - "tools/dockerfreeze"
+
+jobs:
+  # Linting
+  lint:
+    runs-on: ubuntu-latest
+    container: debian:stable
+    steps:
+      - name: Installing dependencies..
+        run: |
+          # Sync repos
+          # Check for git
+          if ! apt list --installed 2>/dev/null | grep -qP "^git\/stable.*"; then
+            # Check if we can install git
+            if ! apt list | grep -qP "^git\/stable.*"; then
+              apt update
+            elif apt list | grep -qP "^git\/stable.*"; then
+              true
+            else
+              exit 255
+            fi
+            # Install git
+            apt install -y git
+          elif apt list --installed 2>/dev/null | grep -qP "^git\/stable.*"; then
+            true
+          else
+            exit 255
+          fi 
+          # Check for shellcheck
+          if ! apt list --installed 2>/dev/null | grep -qP "^shellcheck\s{1}-.*"; then
+            # Check if we can install shellcheck
+            if ! apt list | grep -qP "^shellcheck\s{1}-.*"; then
+              apt update
+            elif apt list | grep -qP "^shellcheck\s{1}-.*"; then
+              true
+            else
+              exit 255
+            fi
+            # Install shellcheck
+            apt install -y shellcheck
+          elif apt list --installed 2>/dev/null | grep -qP "^shellcheck\s{1}-.*"; then
+            true
+          else
+            exit 255
+          fi 
```

```suggestion
          apt update
          apt install git shellcheck
```
I don't think all this code is necessary and makes everything really really complicated.

---

Your method also works, but i like to sanitize for enforcing code quality and optimize my code assuming that `apt update` based on my benchmarks takes additional 2.4 seconds to process not to mension utilizing network resources when it may not be needed.

---

See

```console
kreyren@leonid:~$ benchmark=0; for tries in {1..5}; do SECONDS=0; sudo apt update &>/dev/null; benchmark="$(( benchmark + SECONDS))"; done ; printf 'Action apt update took %s seconds on average\n' "$(( benchmark / 5 ))"
Action apt update took 6 seconds on average

kreyren@leonid:~$ benchmark=0; for tries in {1..5}; do SECONDS=0; if apt list --installed 2>/dev/null | grep -qP "^shellcheck\/stable.*"; then true; fi; benchmark="$(( benchmark + SECONDS))"; done ; printf 'Alternative checking for apt update took %s seconds on average\n' "$(( benchmark / 5 ))"
Alternative checking for apt update took 0 seconds on average
```

---

Honestly those 2.4 seconds are not worth it but I just took 34 lines and converted them to TWO. Readability wins here big big times and your code gets a zero in readability compared to my two lines.

Also there is almost a 0% chance that shellcheck isn't available in the current or next release of either Ubuntu or Debian. And if this would happen it would be almost certain a very short fix and everyone could understand what the code does and does not need to study 34 lines of pretty useless shell code.

---

Mind you this is on system without cached packages and lists ( /var/lib/apt/lists/ and /var/cache/apt/archives ) this would take the additional 0.791 seconds for the required apt update due to the logic checking.

```console
kreyren@leonid:~$ time if ! apt list --installed 2>/dev/null | grep -qP "^shellcehck\s{1}-.*"; then true; fi

real	0m0.791s
user	0m0.713s
sys	0m0.055s
```

---

Furthermore the apt list usage may be optimized by using `apt list shellcheck` instead of current which seems to take less time, but it needs more testing.

Relevant codeblock
```
 if ! apt list | grep -qP "^shellcheck\s{1}-.*"; then
  apt update
elif apt list | grep -qP "^shellcheck\s{1}-.*"; then
  true
else
  exit 255
fi
```

---

FWIW used system for benchmark:

```
kreyren@leonid:~$ neofetch --off
kreyren@leonid 
-------------- 
OS: Kreyrock Linux 9999 Anubis x86_64 
Host: 20089 Lenovo G770 
Kernel: 5.4.0-4-amd64
Uptime: 3 days, 21 hours, 23 mins 
Packages: 2418 (dpkg), 266 (emerge), 339 (cave) 
Shell: bash 5.0.16 
Resolution: 1920x1080 
DE: Xfce 
WM: Xfwm4 
WM Theme: Default 
Theme: Arc-BLACK [GTK2], Adwaita [GTK3] 
Icons: Tango [GTK2], Adwaita [GTK3] 
Terminal: xfce4-terminal 
Terminal Font: Monospace 12 
CPU: Intel i7-2620M (4) @ 3.400GHz 
GPU: AMD ATI Radeon HD 6630M/6650M/6750M/7670M/7690M 
GPU: Intel 2nd Generation Core Processor Family 
Memory: 4721MiB / 7973MiB 
```

---

There is zero reason to do any logic checking. This needs shellcheck in a docker container and the fastest way to do this is to ``apt update`` and ``apt install -y git shellcheck``.

Opsi. I missed the ``-y``.

---

Line 34 is `elif apt list --installed 2>/dev/null | grep -qP "^git\/stable.*"; then`

meaning if git is installed it won't do anything?

> This needs shellcheck in a docker container and the fastest way to do this is to apt update and apt install -y git shellcheck.

True, in mensioned environment your method is faster by around two seconds on average 

First run:
My method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600206205

Your method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600206215

Second run:
My method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600209504

Your method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600206215

Third run:
My method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600211447

Your method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600206215

Fourth run:
My method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600211447

Your method: https://github.com/Kreyren/docker-shellcheck/pull/1/checks?check_run_id=600213018

Which does not necessary make it better for provided reasoning -> I don't want to change this, but you are free to propose your merge request for upstream to decide.

---

There is no need to do anything if git is already installed:
``
time apt install git
Reading package lists... Done
Building dependency tree
Reading state information... Done
git is already the newest version (1:2.20.1-2ubuntu1.19.10.2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

real    0m0.802s
user    0m0.001s
sys     0m0.017s
``

Also we don't care at all about if it is marked as auto installed or manually installed so this is totally fine.